internal/driver/basenotificationresthandler.go

```diff
@@ -0,0 +1,148 @@
+}
+
+func (handler RestNotificationHandler) addContext(next func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	// Add the context with the handler so the endpoint handling code can get back to this handler
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(r.Context(), handlerContextKey, handler)
+		next(w, r.WithContext(ctx))
+	})
+}
+
+// processAsyncRequest receives notification from Onvif camera and sends to the async reading channel
+func (handler RestNotificationHandler) processAsyncRequest(writer http.ResponseWriter, request *http.Request) {
+	vars := mux.Vars(request)
+	deviceName := vars[common.DeviceName]
+	resourceName := vars[common.ResourceName]
+
+	handler.logger.Debugf("Received POST for Device=%s Resource=%s", deviceName, resourceName)
+
+	_, err := handler.service.GetDeviceByName(deviceName)
+	if err != nil {
+		handler.logger.Errorf("Incoming reading ignored. Device '%s' not found", deviceName)
+		http.Error(writer, fmt.Sprintf("Device '%s' not found", deviceName), http.StatusBadRequest)
+		return
+	}
+
+	deviceResource, ok := handler.service.DeviceResource(deviceName, resourceName)
+	if !ok {
+		handler.logger.Errorf("Incoming reading ignored. Resource '%s' not found", resourceName)
+		http.Error(writer, fmt.Sprintf("Resource '%s' not found", resourceName), http.StatusBadRequest)
+		return
+	}
+
+	data, err := handler.readBody(request)
+	if err != nil {
+		handler.logger.Errorf("Incoming reading ignored. Unable to read request body: %s", err.Error())
+		http.Error(writer, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	notify := &event.Notify{}
+	responseEnvelope := gosoap.NewSOAPEnvelope(notify)
+	err = xml.Unmarshal(data, responseEnvelope)
+	if err != nil {
+		handler.logger.Errorf("Failed to create to the subscribe response for Device=%s Resource=%s, %s", deviceName, resourceName, err.Error())
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	cv, err := sdkModel.NewCommandValue(deviceResource.Name, common.ValueTypeObject, notify)
+	if err != nil {
+		handler.logger.Errorf("Failed to create to the commandValue for Device=%s Resource=%s, %s", deviceName, resourceName, err.Error())
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	asyncValues := &models.AsyncValues{
+		DeviceName:    deviceName,
+		CommandValues: []*models.CommandValue{cv},
+	}
+
+	handler.logger.Debugf("Incoming reading received: Device=%s Resource=%s", deviceName, resourceName)
+
+	handler.asyncValues <- asyncValues
+}
+
+func (handler RestNotificationHandler) readBody(request *http.Request) ([]byte, error) {
+	defer request.Body.Close()
+	body, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("no request body provided")
+	}
+
+	return body, nil
+}
+
+func deviceHandler(writer http.ResponseWriter, request *http.Request) {
```

Comment:
Remove, it should never be needed per other comment

Reply:
Removed.